### DISCUSSION TITLE
Achieve persistence after restart

### DISCUSSION THREAD

**Discussion by pvagner** [savonet/liquidsoap#3920] [User Support]
Hello,
I find the tutorials on the web site and these discussions very usefull.
There is a thing I fail to understand and I'd appreciate some hints.

I have prepared a static background track to play if there is nothing else to play.
Then I have music playlists where the order of tracks is random. And specialized music playlists generated by integrating beets library as suggested within the cook book on the website, so I can have 90s hour monday, wednesday and friday and 80s hour tuesday and thursday at 14 PM.
Then there are live shows that are broadcasted as soon as source clients connects. This is all I have understood and the script is clear.

Now I need to figure out how to handle schedule changes and prerecorded shows.
I have a prerecorded show that should play each day at 7:30 pm. Now I do have recorded some 20 episodes of that show and I have a playlist where each line contains anotate protocol pointing to those episodes. Each episode is almost 30 minutes, sometimes it's shorter thus I'd like to fill the rest of those 30 minutes with playing pop music.
The corresponding code is like this:
```
jingles = playlist("~/radio/playlists/jingles.m3u")
pop_music = beets("pop_music", "genre:pop , genre:top , genre:rnb , genre:rock , genre:old length:..4:30")
series_show = playlist(prefix='tmp:', mode='normal',reload=1,reload_mode="rounds", "~/radio/playlists/series.m3u8")
series_show = rotate(weights = [1, 1], [jingles, series_show])
series_show = fallback(track_sensitive = true, [series_show, pop_music])
```

The issue I don't yet have an idea on how to overcome is that when restarting liquidsoap after editing the script the series_shows starts playing from the first item in the playlist eventhough one or more items from that playlist have already played before I have restarted liquidsoap. Is there a way to persist playlist position accross restarts? What is the best strategy here? Or do I have to make a script outside liquidsoap and use request.dynamic to do it?

Also a follow up question I do have a single switch operator in my script listing shows schedule. Is there a way to update this dynamically somehow? As now I have to restart liquidsoap when adding new shows.

Thanks for your advices.

Greetings

Peter

**Comment by vitoyucepi** [comment]
Hi @pvagner,

Probably the best way to achieve persistence is to
1. Use [`playlog`](https://www.liquidsoap.info/doc-2.2.4/reference-extras.html#playlog).
   Also read about it in [the book](https://www.liquidsoap.info/book/book.pdf) on page 84.
1. Implement persistent storage yourself.
   This might be a good solution if `playlog` is not enough.
1. Pull from external service/database with `request.dynamic`.
   It is a good idea to offload the heavy lifting to an external script if you are not very familiar with liquidsoap scripting or if it is missing a feature.
1. Push requests from external script.
   Same as before, but you also offload the timetable functions.
